p2p/protocol/head/head.go

```diff
@@ -0,0 +1,107 @@
+package head
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"net/http"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+)
+
+var log = logging.Logger("go-legs-head")
+
+func deriveProtocolID(topic string) protocol.ID {
+	return protocol.ID("/legs/head/" + topic + "/0.0.1")
+}
+
+func (p *Publisher) Serve(ctx context.Context, host host.Host, topic string) error {
```

Comment:
Passing in a context here is confusing when there is a `Close()` function.  To shutdown the publisher, should the context be canceled, or should `Close` be called?  Can we just have one or the other?

I would prefer that we only use contexts for the cancellation of a function call that may block, not for controlling the lifetime of a service.